sdk/core/core-rest-pipeline/src/restError.ts

```diff
@@ -72,8 +72,8 @@ export class RestError extends Error {
     this.name = "RestError";
     this.code = options.code;
     this.statusCode = options.statusCode;
-    this.request = options.request;
-    this.response = options.response;
+    this.request = options.request && errorSanitizer.sanitizePipelineRequest(options.request);
```

Comment:
I feel like this is changing public surface, since not only are we breaking reference equality of these properties, but also obscuring header and url values where they would have been previously accessible to code. 

If I want to debug a live service issue and I need full unredacted headers to do so, how would I retrieve them after this change goes in?

Reply:
> I feel like this is changing public surface

It is. No argument there. I just don't know how we would fully patch this issue (which has been raised by multiple customers and we have encountered in our own testing) without changing public surface _somehow_. Doing something smaller like updating `toJSON` also carries some risk and does not completely solve the problem. I was hoping that opening this PR would start some discussion and maybe we can come up with a completely different approach.

If we think a different solution would better balance the breaking change risk against the security risk that exists currently then I would be more than happy to close this PR and do something else. I guess I can see a few options:
- We decide that the security risk of the existing behavior is low, so we do nothing and keep the existing behavior. The issue with vitest leaking secrets in our pipelines through `RestError` should be fixed when we upgrade to v2.
- We find a middle-ground solution, like overriding `toJSON`, which will hopefully improve the situation in most cases, but not necessarily all. (if folks have any other middle ground ideas I am all ears!)
- We do what's in this PR and sanitize `RestError`.

> If I want to debug a live service issue and I need full unredacted headers to do so, how would I retrieve them after this change goes in?

The `onResponse` callback would still be available and would provide the unredacted request and response, for whatever that's worth.